File: src/main/java/bot/tree/Node.java
package bot.tree;

import model.procedure.ConsoleOutput;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.function.Consumer;
import java.util.function.IntFunction;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;
import java.util.stream.Stream;

/**
 * A node in the {@link Tree tree-graph}
 */
public class Node implements List<Node> {
    /**
     * An integer associated with that node
     */
    private int val;
    /**
     * A list of children
     */
    private List<Node> children = new ArrayList<>();
    /**
     * a reference to the containing tree
     */
    private Tree tree;
    /**
     * If a Node represents an illegal state of the game it should be invisible meaning no matter what comparison is happening this node is losing
     */
    public boolean invisible;

    /**
     * Constructor
     * @param tree containing tree
     */
    public Node(Tree tree) {
        this(tree, 0);
    }

    /**
     * Constructor
     * @param tree containing tree
     * @param value associated value
     */
    private Node(Tree tree, int value) {
        this.val = value;
        this.tree = tree;
    }

    /**
     * Adds a child to a tree
     */
    void addChild() {
        if (children.isEmpty()) tree.removeLeave(this);
        Node child = new Node(tree);
        children.add(child);
        tree.addLeave(child);
    }

    /**
     * removes a child from this nodes children list
     */
    public void removeChild(Node child) {
        tree.removeLeave(child);
        children.remove(child);
        if (children.isEmpty()) tree.addLeave(this);
    }

    /**
     * gets a child with a given index
     * @param index the index
     * @return the child with given index
     */
    public Node getChild(int index) {
        return children.get(index);
    }

    /**
     * sets the value
     */
    public void setValue(int newValue) {
        val = newValue;
    }

    /**
     * return the value
     */
    public int getValue() {
        return val;
    }

    /**
     * sets this nodes value to the minimum value of all it's child's values
     */
    private void setValueToMinimumChild() {
        val = Integer.MAX_VALUE;
        for (Node child : children) if (child.isVisible() && child.getValue() < val) val = child.getValue();
    }

    /**
     * sets this nodes value to the maximum value of all it's child's values
     */
    private void setValueToMaximumChild() {
        val = Integer.MIN_VALUE;
        for (Node child : children) if (child.isVisible() && child.getValue() > val) val = child.getValue();
    }

    /**
     * recursive method that is invoked alternating with {@link #passLow()}. This method chooses the maximum value of all the children's values and sets it's own value
     * to this value. If their children aren't leaves {@link #passLow()} is invoked on them. Overall this is a simulation of a game where two players traverse
     * a tree and one of them tries to reach leaves with minimal values and the other one wants to reach leaves with maximum values.
     */
    private void passHigh() {
        for (Node child : children) {
            if (child.hasChildren()){
                child.passLow();
            }
        }
        setValueToMaximumChild();
    }

    /**
     * recursive method that is invoked alternating with {@link #passHigh()}. This method chooses the minimum value of all the children's values and sets it's own value
     * to this value. If their children aren't leaves {@link #passHigh()} is invoked on them. Overall this is a simulation of a game where two players traverse
     * a tree and one of them tries to reach leaves with minimal values and the other one wants to reach leaves with maximum values.
     */
    private void passLow() {
        for (Node child : children) {
            if (child.hasChildren()){
                child.passHigh();
            }
        }
        setValueToMinimumChild();
    }

    /**
     * This Method uses {@link #passLow()} to determine the index of the child with the best outcome for high values in a game where two players traverse
     * a tree and one of them tries to reach leaves with minimal values and the other one wants to reach leaves with maximum values.
     * @return the index of the best child
     */
    public int indexOfNodeWithBestExpectationOfHighValue() {
        int bestValue = Integer.MIN_VALUE;
        List<Integer> bestIndices = new ArrayList<>();
        for (int i = 0; i < children.size(); i++) {
            if (children.get(i).isVisible()) {
                children.get(i).passLow();
                if (ConsoleOutput.treeTraversal) System.out.println("Dir: " + i + ", Rating: " + children.get(i).getValue());
                if (children.get(i).getValue() > bestValue) {
                    bestValue = children.get(i).getValue();
                    bestIndices.clear();
                    bestIndices.add(i);
                } else if (children.get(i).getValue() == bestValue) {
                    bestIndices.add(i);
                }
            }
        }
        val = bestValue;
        return bestIndices.get((int)(Math.random() * bestIndices.size()));
    }

    /**
     * return true if this node has children, false otherwise
     * @return true if this node has children, false otherwise
     */
    private boolean hasChildren() {
        return !children.isEmpty();
    }

    /**
     * deletes all children nodes
     */
    public void makeLeave() {
        if (children.isEmpty()) return;
        for (int i = 0; i < tree.getAmountOfLeaves(); i++) {
            if (isParentOf(tree.getLeave(i))) tree.removeLeave(tree.getLeave(i));
        }
        children = new ArrayList<>();
        tree.addLeave(this);
    }

    /**
     * returns true if this node is a (grand)*parent of the given node
     * @param other other node
     */
    private boolean isParentOf(Node other) {
        for (Node child : children) if (child.equals(other)) return true;
        for (Node child : children) if (child.isParentOf(other)) return true;
        return false;
    }

    @Override
    public String toString() {
        return recToString(0);
    }

    private String recToString(int spaceOffset) {
        StringBuilder str = new StringBuilder();
        for(int i = 0; i < spaceOffset; i++) str.append(" ");
        str.append(val).append("\n");
        for (Node child : children) {
            str.append(child.recToString(spaceOffset + 4));
        }
        return str.toString();
    }

    public void setInvisible() {
        invisible = true;
    }

    private boolean isVisible() {
        return !invisible;
    }

    boolean isLeave() {
        return children.size() == 0;
    }

    int getAmountOfChildren() {
        return children.size();
    }

    @Override
    public int size() {
        return children.size();
    }

    @Override
    public boolean isEmpty() {
        return children.isEmpty();
    }

    @Override
    public boolean contains(Object o) {
        return children.contains(o);
    }

    @NotNull
    @Override
    public Iterator<Node> iterator() {
        return children.iterator();
    }

    @Override
    public void forEach(Consumer<? super Node> action) {
        children.forEach(action);
    }

    @NotNull
    @Override
    public Object[] toArray() {
        return children.toArray();
    }

    @NotNull
    @Override
    public <T> T[] toArray(@NotNull T[] ts) {
        return children.toArray(ts);
    }

    @Override
    public <T> T[] toArray(IntFunction<T[]> generator) {
        return children.toArray(generator);
    }

    @Override
    public boolean add(Node node) {
        return children.add(node);
    }

    @Override
    public boolean remove(Object o) {
        return children.remove(o);
    }

    @Override
    public boolean containsAll(@NotNull Collection<?> collection) {
        return children.containsAll(collection);
    }

    @Override
    public boolean addAll(@NotNull Collection<? extends Node> collection) {
        return children.addAll(collection);
    }

    @Override
    public boolean addAll(int i, @NotNull Collection<? extends Node> collection) {
        return children.addAll(i, collection);
    }

    @Override
    public boolean removeAll(@NotNull Collection<?> collection) {
        return children.removeAll(collection);
    }

    @Override
    public boolean removeIf(Predicate<? super Node> filter) {
        return children.removeIf(filter);
    }

    @Override
    public boolean retainAll(@NotNull Collection<?> collection) {
        return children.retainAll(collection);
    }

    @Override
    public void replaceAll(UnaryOperator<Node> operator) {
        children.replaceAll(operator);
    }

    @Override
    public void sort(Comparator<? super Node> c) {
        children.sort(c);
    }

    @Override
    public void clear() {
        children.clear();
    }

    @Override
    public Node get(int i) {
        return children.get(i);
    }

    @Override
    public Node set(int i, Node node) {
        return children.set(i, node);
    }

    @Override
    public void add(int i, Node node) {
        children.add(i, node);
    }

    @Override
    public Node remove(int i) {
        return children.remove(i);
    }

    @Override
    public int indexOf(Object o) {
        return children.indexOf(o);
    }

    @Override
    public int lastIndexOf(Object o) {
        return children.lastIndexOf(o);
    }

    @NotNull
    @Override
    public ListIterator<Node> listIterator() {
        return children.listIterator();
    }

    @NotNull
    @Override
    public ListIterator<Node> listIterator(int i) {
        return children.listIterator(i);
    }

    @NotNull
    @Override
    public List<Node> subList(int i, int i1) {
        return children.subList(i, i1);
    }

    @Override
    public Spliterator<Node> spliterator() {
        return children.spliterator();
    }

    @Override
    public Stream<Node> stream() {
        return children.stream();
    }

    @Override
    public Stream<Node> parallelStream() {
        return children.parallelStream();
    }
}

File: src/main/java/bot/bots/tree/Node.kt
package bot.bots.tree

import model.procedure.ConsoleOutput
import java.util.*
import java.util.function.Consumer
import java.util.function.Predicate
import java.util.function.UnaryOperator
import java.util.stream.Stream

/**
 * A node in the [tree-graph][Tree]
 */
open class Node(/*var tree: Tree<Node>? = null, */val children: MutableList<Node> = ArrayList(), var value: Int = 0) :
    MutableList<Node> {

    @Deprecated("prune tree instead", ReplaceWith("nothing"))
    private val isVisible: Boolean
        get() = !invisible
    val isLeaf: Boolean
        get() = children.size == 0

    var parent: Node = this

    val amountDescendants: Int
        get() {
            return children.size + children.sumOf { it.amountDescendants }
        }

    /**
     * If a Node represents an illegal state of the game it should be invisible meaning no matter what comparison is happening this node is losing
     */
    @Deprecated("prune tree instead")
    var invisible = false

    /* /**
      * Adds a child to a tree
      */
     fun addChild(child: Node) {
         if (children.isEmpty()) tree!!.leaves.remove(this)
         children.add(child)
         tree!!.addLeave(child)
     }

     /**
      * removes a child from this nodes children list
      */
     fun removeChild(child: Node) {
         tree!!.leaves.remove(child)
         children.remove(child)
         if (children.isEmpty()) tree!!.addLeave(this)
     }*/

    /**
     * sets this nodes value to the minimum value of all it's child's values
     */
    private fun setValueToMinimumChild() {
        value = Int.MAX_VALUE
        for (child in children) if (child.isVisible && child.value < value) value = child.value
    }

    /**
     * sets this nodes value to the maximum value of all it's child's values
     */
    private fun setValueToMaximumChild() {
        value = Int.MIN_VALUE
        for (child in children) if (child.isVisible && child.value > value) value = child.value
    }

    /**
     * recursive method that is invoked alternating with [.passLow]. This method chooses the maximum value of all the children's values and sets it's own value
     * to this value. If their children aren't leaves [.passLow] is invoked on them. Overall this is a simulation of a game where two players traverse
     * a tree and one of them tries to reach leaves with minimal values and the other one wants to reach leaves with maximum values.
     */
    private fun passHigh() {
        for (child in children) {
            if (child.hasChildren()) {
                child.passLow()
            }
        }
        setValueToMaximumChild()
    }

    /**
     * recursive method that is invoked alternating with [.passHigh]. This method chooses the minimum value of all the children's values and sets it's own value
     * to this value. If their children aren't leaves [.passHigh] is invoked on them. Overall this is a simulation of a game where two players traverse
     * a tree and one of them tries to reach leaves with minimal values and the other one wants to reach leaves with maximum values.
     */
    private fun passLow() {
        for (child in children) {
            if (child.hasChildren()) {
                child.passHigh()
            }
        }
        setValueToMinimumChild()
    }

    /**
     * This Method uses [.passLow] to determine the index of the child with the best outcome for high values in a game where two players traverse
     * a tree and one of them tries to reach leaves with minimal values and the other one wants to reach leaves with maximum values.
     * @return the index of the best child
     */
    fun indexOfNodeWithBestExpectationOfHighValue(): Int {
        var bestValue = Int.MIN_VALUE
        val bestIndices: MutableList<Int> = ArrayList()
        for (i in children.indices) {
            if (children[i].isVisible) {
                children[i].passLow()
                if (ConsoleOutput.treeTraversal) println("Dir: " + i + ", Rating: " + children[i].value)
                if (children[i].value > bestValue) {
                    bestValue = children[i].value
                    bestIndices.clear()
                    bestIndices.add(i)
                } else if (children[i].value == bestValue) {
                    bestIndices.add(i)
                }
            }
        }
        value = bestValue
        return bestIndices[(Math.random() * bestIndices.size).toInt()]
    }

    /**
     * return true if this node has children, false otherwise
     * @return true if this node has children, false otherwise
     */
    private fun hasChildren(): Boolean {
        return children.isNotEmpty()
    }

    /*/**
     * deletes all children nodes
     */
    fun makeLeave() {
        if (children.isEmpty()) return
        for (leave in tree!!.leaves) {
            if (isParentOf(leave)) tree!!.leaves.remove(leave)
        }
        children.clear()
        tree!!.addLeave(this)
    }*/

    /**
     * returns true if this node is a (grand)*parent of the given node
     * @param other other node
     */
    fun isParentOf(other: Node): Boolean {
        return other.isDescendantOf(this)
           // for (child in children) if (child == other) return true
            //for (child in children) if (child.isParentOf(other)) return true
            //return false
    }

    /**
     * returns true if this node is a (grand)*child of the given node
     * @param other other node
     */
    fun isDescendantOf(other: Node): Boolean {
        if (parent == other) return true
        if (parent == this) return false
        return parent.isDescendantOf(other)
    }

    override fun toString(): String {
        return recToString(0)
    }

    private fun recToString(spaceOffset: Int): String {
        val str = StringBuilder()
        for (i in 0 until spaceOffset) str.append(" ")
        str.append(value).append("\n")
        for (child in children) {
            str.append(child.recToString(spaceOffset + 4))
        }
        return str.toString()
    }

    override val size: Int
        get() = children.size

    override fun isEmpty(): Boolean {
        return children.isEmpty()
    }

    override operator fun contains(element: Node): Boolean {
        return children.contains(element)
    }

    override fun iterator(): MutableIterator<Node> {
        return children.iterator()
    }

    override fun forEach(action: Consumer<in Node>) {
        children.forEach(action)
    }

    override fun add(element: Node): Boolean {
        return children.add(element)
    }

    override fun remove(element: Node): Boolean {
        return children.remove(element)
    }

    override fun containsAll(elements: Collection<Node>): Boolean {
        return children.containsAll(elements)
    }

    override fun addAll(elements: Collection<Node>): Boolean {
        return children.addAll(elements)
    }

    override fun addAll(index: Int, elements: Collection<Node>): Boolean {
        return children.addAll(index, elements)
    }

    override fun removeAll(elements: Collection<Node>): Boolean {
        return children.removeAll(elements)
    }

    override fun removeIf(filter: Predicate<in Node>): Boolean {
        return children.removeIf(filter)
    }

    override fun retainAll(elements: Collection<Node>): Boolean {
        return children.retainAll(elements)
    }

    override fun replaceAll(operator: UnaryOperator<Node>) {
        children.replaceAll(operator)
    }

    override fun sort(c: Comparator<in Node>) {
        children.sortWith(c)
    }

    override fun clear() {
        children.clear()
    }

    override fun get(index: Int): Node {
        return children[index]
    }

    override fun set(index: Int, element: Node): Node {
        return children.set(index, element)
    }

    override fun add(index: Int, element: Node) {
        children.add(index, element)
    }

    override fun removeAt(index: Int): Node {
        return children.removeAt(index)
    }

    override fun indexOf(element: Node): Int {
        return children.indexOf(element)
    }

    override fun lastIndexOf(element: Node): Int {
        return children.lastIndexOf(element)
    }

    override fun listIterator(): MutableListIterator<Node> {
        return children.listIterator()
    }

    override fun listIterator(index: Int): MutableListIterator<Node> {
        return children.listIterator(index)
    }

    override fun subList(fromIndex: Int, toIndex: Int): MutableList<Node> {
        return children.subList(fromIndex, toIndex)
    }

    override fun spliterator(): Spliterator<Node> {
        return children.spliterator()
    }

    override fun stream(): Stream<Node> {
        return children.stream()
    }

    override fun parallelStream(): Stream<Node> {
        return children.parallelStream()
    }
}